File: julia/recursion.jl
using BenchmarkTools;

function factorial(n::Int64) ::Int64
    if n == 0
        return 1;
    end
    return n*factorial(n-1);
end

function fibonacci(n::Int64) ::Int64
    if n <= 1
        return 1;
    end
    return fibonacci(n-1) + fibonacci(n-2);
end

function measure_factorial()
    list::Vector{Int64} = [x for x in 1000:2000-1];
    exec_times::Vector{Float64} = zeros(1000);
    for i in 1:length(list)
        item = list[i];
        result = @timed factorial(item);
        exec_times[i] = result.time;
    end
    println(median(exec_times));

end


function measure_fibonacci()
    n::Int64 = 40;
    result = @timed fibonacci(n);
    println(result.time);
end

function main()
    measure_factorial();
    measure_fibonacci();

end

main()

File: python/recursion.py
import sys
import time
from statistics import median

sys.setrecursionlimit(5000)

def count_time(start=0):
    return time.perf_counter() - start


def factorial(n):
    if n == 0:
        return 1
    return n*factorial(n-1)


def fibonacci(n):
    if n <= 1:
        return 1
    return fibonacci(n-1) + fibonacci(n-2)


def measure_factorial():
    size = 1000
    list_ = [e for e in range(1000, 1000+size)]
    exec_times = list()
    for integer in list_:
        start = count_time()
        r = factorial(integer)
        counted_time = count_time(start)
        exec_times.append(counted_time)
    print(median(exec_times))


def measure_fibonacci():
    n = 40
    start = count_time()
    result = fibonacci(n)
    counted_time = count_time(start)
    print(counted_time)


def main(): #crashed in integer==2858
    measure_factorial()
    measure_fibonacci()


if __name__ == '__main__':
    main()